data/libs/EquipType.lua

```diff
@@ -325,17 +431,90 @@ function HyperdriveType:OnLeaveHyperspace(ship)
 	end
 end
 
+--==============================================================================
+
 -- NOTE: "sensors" have no general-purpose code associated with the equipment type
-local SensorType = utils.inherits(EquipType, "SensorType")
+---@class Equipment.SensorType : EquipType
+local SensorType = utils.inherits(EquipType, "Equipment.SensorType")
+
+--==============================================================================
 
 -- NOTE: all code related to managing a body scanner is implemented in the ScanManager component
-local BodyScannerType = utils.inherits(SensorType, "BodyScannerType")
+---@class Equipment.BodyScannerType : EquipType
+---@field stats table
+local BodyScannerType = utils.inherits(SensorType, "Equipment.BodyScannerType")
+
+--==============================================================================
+
+---@class Equipment.CabinType : EquipType
+---@field passengers Character[]?
+local CabinType = utils.inherits(EquipType, "Equipment.CabinType")
+
+---@param passenger Character
+function CabinType:AddPassenger(passenger)
+	table.insert(self.passengers, passenger)
+	self.icon_name = "equip_cabin_occupied"
+end
+
+---@param passenger Character
+function CabinType:RemovePassenger(passenger)
+	utils.remove_elem(self.passengers, passenger)
+	self.icon_name = "equip_cabin_empty"
+end
+
+function CabinType:HasPassenger(passenger)
+	return utils.contains(self.passengers, passenger)
+end
+
+function CabinType:GetNumPassengers()
+	return self.passengers and #self.passengers or 0
+end
+
+function CabinType:GetMaxPassengers()
+	return self.capabilities.cabin
+end
+
+function CabinType:GetFreeBerths()
+	return self.capabilities.cabin - (self.passengers and #self.passengers or 0)
+end
+
+function CabinType:OnInstall(ship, slot)
+	EquipType.OnInstall(self, ship, slot)
+
+	self.passengers = {}
+end
+
+function CabinType:OnRemove(ship, slot)
+	EquipType.OnRemove(self, ship, slot)
+
+	if self.passengers then
+		logWarning("Removing passenger cabin with passengers onboard!")
+		ship:setprop("cabin_occupied_cap", ship["cabin_occupied_cap"] - #self.passengers)
+	end
```

Comment:
Crash Bug:
```
libs/EquipType.lua:492: attempt to perform arithmetic on field 'cabin_occupied_cap' (a nil value)
```

This is due to "self.passengers" not being NULL.

Also, Q: is self.passengers guaranteed to be indexed by integers? Because if not, "#self.passengers" will not return a correct value. Also, given the code uses the Lua `table` interface, wouldn't it be more consistent to use `table.getn(self.passengers)` ?

Proposed Fix:
```suggestion
	if self:GetNumPassengers() > 0 then
		logWarning("Removing passenger cabin with passengers onboard!")
		ship:setprop("cabin_occupied_cap", ship["cabin_occupied_cap"] - self:GetNumPassengers())
	end
```

Reply:
.. also, there's a log warning message that a cabin with a passenger was sold, but there's nothing preventing the user from (accidentally) actually doing so.

Future improvement: suggest a confirmation popup.

Reply:
`#table` is a Lua fast-path provided to query the length of the "array part" of a table in roughly constant time. It returns the length of the sequence of consecutive non-nil values at integer indices 1..N.

This is a refactoring bug - originally cabins were only allowed to contain one passenger. Will fix.

Existing code had no confirmation popup and allowed selling occupied passenger cabins. I've deferred adding a popup/confirmation/restrictions until a further block of work on "passenger management".